### PR TITLE
Refactor LoRA weights merging + support FSDP

### LINKED ISSUE #161: Support LoRA with multiple devices
When switching from DeepSpeed stage 2 over to DeepSpeed stage 3, there are currently issues with loading the model via the LoRA finetuning script:

```
...
 18176]) from checkpoint, the shape in current model is torch.Size([0]).
        size mismatch for transformer.h.31.norm_1.weight: copying a param with shape torch.Size([4544]) from checkpoint, the shape in current model is torch.Size([0]).
        size mismatch for transformer.h.31.norm_1.bias: copying a param with shape torch.Size([4544]) from checkpoint, the shape in current model is torch.Size([0]).
        size mismatch for transformer.h.31.attn.attn.weight: copying a param with shape torch.Size([4672, 4544]) from checkpoint, the shape in current model is torch.Size([0]).
        size mismatch for transformer.h.31.attn.proj.weight: copying a param with shape torch.Size([4544, 4544]) from checkpoint, the shape in current model is torch.Size([0]).
        size mismatch for transformer.h.31.mlp.fc.weight: copying a param with shape torch.Size([18176, 4544]) from checkpoint, the shape in current model is torch.Size([0]).
        size mismatch for transformer.h.31.mlp.proj.weight: copying a param with shape torch.Size([4544, 18176]) from checkpoint, the shape in current model is torch.Size([0]).
        size mismatch for transformer.ln_f.weight: copying a param with shape torch.Size([4544]) from checkpoint, the shape in current model is torch.Size([0]).
        size mismatch for transformer.ln_f.bias: copying a param with shape torch.Size([4544]) from checkpoint, the shape in current model is torch.Size([0]).
```

But I see that an FSDP update is in the works via  #118, so I am not sure if that's worth addressing right now.

### PR DESCRIPTION
**Pull request by awaelchli**
Fixes #161

Refactors the merging of LoRA weights.

**Before:**
Merging was implemented in `model.eval()`, and "unmerging" was implemented in `model.train()`. This was done to speed up the forward pass for inference where LoRA matrices aren't needed. Unmerging was needed to bring the weights back for optimization (assuming validation is interleaved at periodic intervals with training). However, due to the fact that the `eval()/train()` methods access the weights, they are sharded and so computations on them can't be performed directly. In FSDP, the full weights are only available during the forward. 

**Now:**
We don't need merging and unmerging implemented in `eval()/train()`. Our validation loop will be slightly slower but we can afford that cost since we're not validating extensively. The impact is not noticeable. Hence, we can move the implementation to a separate `merge()` method that can be used outside of training, e.g. when performing inference (see `generate/lora.py`).
Thus, `eval()` and `train()` are now used in the normal way without merging code, and are compatible with FSDP.

Notes:
1. Verified loss is decreasing, no negative impacts on convergence (1b pythia models, 4 devices, trained until loss < 1.0)
2. Instantiation is slow because model gets created in CPU memory first. This is expected because we don't yet have a reliable fast implementation for init-module with FSDP.

### CODE REVIEW

File: lit_gpt/lora.py [hunk 1]
```diff
@@ -354,36 +335,17 @@ def zero_pad(self, x: torch.Tensor) -> torch.Tensor:
         result = result.index_copy(1, self.lora_ind, x.reshape(-1, shape))  # (4096, 256)
         return result.view((*x.shape[:-1], self.out_features)).transpose(0, 1)  # (64, 64, 384)
 
-    def train(self, mode: bool = True):
```
Comment (awaelchli) [inline lit_gpt/lora.py:357]: The previous code supported "unmerging" too, but we don't need it. If you want to keep it nonetheless (e.g. in a unmerge method), let me know.

File: finetune/lora.py [hunk 2]
```diff
@@ -24,7 +24,7 @@
 save_interval = 100
 eval_iters = 100
 log_interval = 1
-devices = 1
+devices = 2
```
Comment (awaelchli) [inline finetune/lora.py:27]: todo undo hparam changes

File: finetune/lora.py [hunk 3]
```diff
@@ -126,6 +132,11 @@ def main(fabric: L.Fabric, data_dir: Path, checkpoint_dir: Path, out_dir: Path):
     fabric.seed_everything(1337 + fabric.global_rank)
 
     train_time = time.time()
+    
+    # sanity check
+    save_path = out_dir / "sanity_checkpoint.pth"
+    save_lora_checkpoint(fabric, model, save_path)
+
```
Comment (awaelchli) [inline finetune/lora.py:139]:
```suggestion
```

File: lit_gpt/lora.py [hunk 4]
```diff
@@ -307,7 +288,7 @@ def __init__(
                 lora_ind.append(
                     torch.arange(self.in_features + self.kv_embd_size, self.out_features, device=self.weight.device)
                 )
-            self.lora_ind = torch.cat(lora_ind)
+            self.register_buffer("lora_ind", torch.cat(lora_ind))
```
[the remark this answers is not in the document]
Comment (carmocca) [reply lit_gpt/lora.py:291]: Should this set `persistent=False` so it's not included in the state dict?

File: generate/lora.py [hunk 5]
```diff
@@ -111,6 +111,7 @@ def main(
     fabric.print(f"Time to load the model weights: {time.time() - t0:.02f} seconds.", file=sys.stderr)
 
     model.eval()
+    merge_lora_weights(model)
```
Comment (carmocca) [inline generate/lora.py:114]: Is this in case the weights haven't been merged already?

What's more common, checkpoints with the weights merged or unmerged?

File: lit_gpt/lora.py [hunk 6]
```diff
@@ -710,3 +671,11 @@ def __init__(self, config: Config) -> None:
             lora_alpha=config.alpha,
             lora_dropout=config.dropout,
         )
+
+
+def merge_lora_weights(model: GPT) -> None:
+    """Merge LoRA weights into the full-rank weights to speed up inference."""
+    for module in model.modules():
+        merge_fn = getattr(module, "merge", None)
+        if callable(merge_fn):
```
Comment (carmocca) [inline lit_gpt/lora.py:680]: Why not use `isinstance(module, LoRALinear)`?

File: lit_gpt/lora.py [hunk 7]
```diff
@@ -482,7 +443,7 @@ def mark_only_lora_as_trainable(model: nn.Module, bias: str = "none") -> None:
 
 
 def lora_filter(key: str, value: Any) -> bool:
-    return "lora_" in key
+    return "lora_" in key and "lora_ind" not in key
```
Comment (carmocca) [inline lit_gpt/lora.py:446]: If it's made persistent then this wouldn't be necessary

**Review by carmocca — APPROVED** [review]
:100:

File: generate/lora.py (hunk 5 above)
Comment (awaelchli) [reply generate/lora.py:114]: Yeah, the checkpoint produced by the lora finetuning scripts only saves the lora weights, meaning they are not merged. This is common, because we can store these in a few MB vs. a full checkpoint (GBs). Additionally, it is common to share the LoRA weights if the original weights had a license that prevented the distribution of checkpoints (LLaMA v1). Plus, the merging is very cheap. So there are only benefits to this.

File: lit_gpt/lora.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply lit_gpt/lora.py]: Done!

File: generate/lora.py (hunk 5 above)
Comment (awaelchli) [reply generate/lora.py:114]: Oh and you see in the line above `model.eval()`, this is where the merging happened previously, so we're not changing  how it was before.

File: lit_gpt/lora.py [hunk 8]
```diff
@@ -151,36 +151,17 @@ def reset_parameters(self):
...
-                if self.r > 0:
-                    self.weight.data -= T(self.lora_B @ self.lora_A) * self.scaling
-                self.merged = False
-        else:
-            if self.merge_weights and not self.merged:
-                # Merge the weights and mark it
-                if self.r > 0:
-                    self.weight.data += T(self.lora_B @ self.lora_A) * self.scaling
-                self.merged = True
+        if self.merge_weights and not self.merged:
```
Comment (Andrei-Aksionov) [inline lit_gpt/lora.py:160]: I feel somewhat uncomfortable about it. 
If someone explicitly calls `merge` method - that doesn't necessarily mean that the weights will be indeed merged:
1. if `self.merge_weights` is False 
2. if `self.r` <= 0

... and it will happen without any warnings.

But also the question is do we need those checks?
1. `self.merge_weights` perhaps was a way to prevent automatically merging weights in `train` method. Since now merging is done explicitly we don't need it.
2. `self.r = -1` is a way to disable LoRA weights and get a regular Linear layer, but we don't use this functionality. For me it look like using `nn.Linear` if we don't need LoRA functionality is simpler (more obvious). But I might be wrong.
So we could instead of:
```python
if config.to_projection:
    self.proj = LoRALinear(
        config.n_embd,
        config.n_embd,
        bias=config.bias,
        r=config.r,
        lora_alpha=config.alpha,
        lora_dropout=config.dropout,
    )
else:
    self.proj = nn.Linear(config.n_embd, config.n_embd, bias=config.bias)
```
use this
```python
self.proj = LoRALinear(
    config.n_embd,
    config.n_embd,
    bias=config.bias,
    r=config.r if config.to_projection else -1,
    lora_alpha=config.alpha,
    lora_dropout=config.dropout,
)
```
---
So I think either at least add some warning information or get rid of `merge_weights` and assert that `self.r` is greater than 0.

What do you think @awaelchli ?
I tend to over-complicate things, so I'm relying on your input.

Comment (awaelchli) [reply lit_gpt/lora.py:160]: I am not concerned about it, we use `merge_weights=True` everywhere so I don't think this can have negative impacts. The code for this was inherited from the original lora implementation: https://github.com/microsoft/LoRA/blob/998cfe4d351f4d6b4a47f0921dec2397aa0b9dfe/loralib/layers.py#L65
If r = 0, that means LoRA is disabled (delta_w = 0) so merging being a no-op is not unreasonable. Whether this should error/be allowed or not is merely a design decision IMO. 
I'm not opposed to removing the merge_weights attribute because it's not useful anymore, and we don't use it in our scripts.

Comment (Andrei-Aksionov) [reply lit_gpt/lora.py:160]: I'll create a PR if you don't mind.

File: finetune/lora.py [hunk 9]
```diff
@@ -66,7 +66,13 @@ def setup(
             fabric_devices = "auto"
             strategy = XLAStrategy(sync_module_states=False)
         else:
-            raise NotImplementedError
+            precision="bf16-true"
```
Comment (carmocca) [inline finetune/lora.py:69]: @awaelchli Did you accidentally leave this hardcoded or is this on purpose?

Comment (awaelchli) [reply finetune/lora.py:69]: bf16-true is what I tested it with yes. We can test other precision settings but for now we know this one works fine.

Comment (carmocca) [reply finetune/lora.py:69]: The script exposes a `precision` argument, so if you'd prefer that we limit the options, it should at least error out if the CLI argument is different
